poetry/core/masonry/builders/builder.py

```diff
@@ -115,58 +131,70 @@ def is_excluded(self, filepath):  # type: (Union[str, Path]) -> bool
+
             for file in include.elements:
                 if "__pycache__" in str(file):
                     continue
 
                 if file.is_dir():
+                    if self.format in formats:
+                        for f in file.glob("**/*"):
+                            rel_path = f.relative_to(self._path)
+
+                            if (
+                                rel_path not in set([t.path for t in to_add])
+                                and not f.is_dir()
+                                and not self.is_excluded(rel_path)
+                            ):
+                                to_add.append(
+                                    IncludeFile(path=rel_path, source_root=self._path)
+                                )
                     continue
 
-                file = file.relative_to(self._path)
+                if (
+                    isinstance(include, PackageInclude)
+                    and include.source
+                    and self.format == "wheel"
+                ):
+                    source_root = include.base
+                else:
+                    source_root = self._path
+
+                file = file.relative_to(source_root)
+                include_file = IncludeFile(path=file, source_root=source_root.resolve())
 
                 if self.is_excluded(file) and isinstance(include, PackageInclude):
                     continue
 
                 if file.suffix == ".pyc":
                     continue
 
-                if file in to_add:
+                if file in set([f.path for f in to_add]):
                     # Skip duplicates
                     continue
 
                 logger.debug(" - Adding: {}".format(str(file)))
-                to_add.append(file)
-
-        # Include project files
-        logger.debug(" - Adding: pyproject.toml")
-        to_add.append(Path("pyproject.toml"))
-
-        # If a license file exists, add it
-        for license_file in self._path.glob("LICENSE*"):
-            logger.debug(" - Adding: {}".format(license_file.relative_to(self._path)))
-            to_add.append(license_file.relative_to(self._path))
-
-        # If a README is specified we need to include it
-        # to avoid errors
-        if "readme" in self._poetry.local_config:
-            readme = self._path / self._poetry.local_config["readme"]
-            if readme.exists():
-                logger.debug(" - Adding: {}".format(readme.relative_to(self._path)))
-                to_add.append(readme.relative_to(self._path))
-
-        # If a build script is specified and explicitely required
+                to_add.append(include_file)
+
+        # If a build script is specified and explicitly required
         # we add it to the list of files
         if self._package.build and not exclude_build:
-            to_add.append(Path(self._package.build))
+            to_add.append(
+                IncludeFile(path=Path(self._package.build), source_root=self._path)
```

Comment:
We can move, the `Path()`. `.relative_to()` and `.resolve()` logic into the class itself I think.